proposals/gc/Post-MVP.md

```diff
@@ -0,0 +1,563 @@
+  )
+
+  (func
+    (typeparam $T)
+    (param $rttT (rtt (typeparam $T)))
+    (param $x (ref $T))
+    (result (ref $Pair (typeparam $T)))
+
+    ;; allocate a pair with full RTT information
+    (struct.new_with_rtt
+      (rtt.canon $Pair (typeparam $T) (local.get $rttT))
+      (local.get $x) (local.gt $x)
+    )
+  )
+  ```
+
+
+## Variants
+
+The MVP supports the most basic form of _pointer tagging_ via the `i31ref` type.
+That type allows injecting and distinguishing unboxed integers and pointers in the same type space, such that unboxing can be guaranteed on all platforms.
+
+However, many language implementations use more elaborate tagging schemes in one form or the other. For example, they want to efficiently distinguish different classes of values without dereferencing, or to store additional information in pointer values without requiring extra space (e.g., Lisp or Prolog often introduce a special tag bit for cons cells).
+Other languages provide user-defined tags as an explicit language feature (e.g., _variants_ or _algebraic data types_).
+
+Unfortunately, hardware differs widely in how many tagging bits a pointer can conveniently support, and different VMs might have additional constraints on how many of these bits they can make available to user code.
+In order to provide tagging in a way that is portable but maximally efficient on any given hardware and engine, a somewhat higher level of abstraction is useful.
+
+Such a more high-level solution would be to support a form of _variant types_ (a.k.a. disjoint unions or sum types) in the type system.
+In addition to structs and arrays, a module could define a variant type that is a closed union of multiple different cases.
+Dedicated instructions allow allocating and inspecting references of variant type.
+
+It is left to the engine to pick an efficient representation for the required tags, and depending on the hardware's word size, the number of tags in a defined type, the presence of parameters to a given tag, and other design decisions in the engine, these tags could either be stored as bits in the pointer, in a shared per-type data structure (a.k. hidden class or shape), or in an explicit per-value slot within the heap object.
+These decisions can be made by the engine on a per-type basis; validation ensures that all uses are coherent.
+
+**Why Post-MVP:** Variants allow for more compact representations and potentially more precise types, thereby possibly saving a cetain number of runtime checks and downcasts over the use of a common supertype, as necessary in the MVP. However, they are not strictly necessary in the presence of the latter. Nor can they replace it, since they necessarily define a _closed_ set of values, whereas the ability to import an arbitrary number of abstract types requires a way to include an _open_ set of types. To handle that, yet another form of _extensible union types_ (with generative tags) would be required in addition.
```

Comment:
```suggestion
**Why Post-MVP:** Variants allow for more compact representations and potentially more precise types, thereby possibly saving a certain number of runtime checks and downcasts over the use of a common supertype, as necessary in the MVP. However, they are not strictly necessary in the presence of the latter. Nor can they replace it, since they necessarily define a _closed_ set of values, whereas the ability to import an arbitrary number of abstract types requires a way to include an _open_ set of types. To handle that, yet another form of _extensible union types_ (with generative tags) would be required in addition.
```

Reply:
The above change just changes "cetain" to "certain."